test-data/unit/check-parameter-specification.test

```diff
@@ -28,51 +28,303 @@ def foo6(x: Callable[[P], int]) -> None: ...  # E: Invalid location for ParamSpe
+        kwargs = o2  # E: Incompatible types in assignment (expression has type "object", variable has type "P.kwargs")
+        a: Any
+        a = args
+        a = kwargs
+        args = kwargs  # E: Incompatible types in assignment (expression has type "P.kwargs", variable has type "P.args")
+        kwargs = args  # E: Incompatible types in assignment (expression has type "P.args", variable has type "P.kwargs")
+        args = a
+        kwargs = a
+[builtins fixtures/dict.pyi]
+
+[case testParamSpecSubtypeChecking2]
+from typing import Callable, Generic
+from typing_extensions import ParamSpec
+
+P = ParamSpec('P')
+P2 = ParamSpec('P2')
+
+class C(Generic[P]):
+    pass
+
+def f(c1: C[P], c2: C[P2]) -> None:
+    c1 = c1
+    c2 = c2
+    c1 = c2  # E: Incompatible types in assignment (expression has type "C[P2]", variable has type "C[P]")
+    c2 = c1  # E: Incompatible types in assignment (expression has type "C[P]", variable has type "C[P2]")
+
+def g(f: Callable[P, None], g: Callable[P2, None]) -> None:
+    f = f
+    g = g
+    f = g  # E: Incompatible types in assignment (expression has type "Callable[P2, None]", variable has type "Callable[P, None]")
+    g = f  # E: Incompatible types in assignment (expression has type "Callable[P, None]", variable has type "Callable[P2, None]")
+[builtins fixtures/dict.pyi]
+
+[case testParamSpecJoin]
+from typing import Callable, Generic, TypeVar
+from typing_extensions import ParamSpec
+
+P = ParamSpec('P')
+P2 = ParamSpec('P2')
+P3 = ParamSpec('P3')
+T = TypeVar('T')
+
+def join(x: T, y: T) -> T: ...
+
+class C(Generic[P, P2]):
+    def m(self, f: Callable[P, None], g: Callable[P2, None]) -> None:
+        reveal_type(join(f, f))  # N: Revealed type is "def (*P.args, **P.kwargs)"
+        reveal_type(join(f, g))  # N: Revealed type is "builtins.function*"
+
+    def m2(self, *args: P.args, **kwargs: P.kwargs) -> None:
+        reveal_type(join(args, args))  # N: Revealed type is "P.args`1"
+        reveal_type(join(kwargs, kwargs))  # N: Revealed type is "P.kwargs`1"
+        reveal_type(join(args, kwargs))  # N: Revealed type is "builtins.object*"
+        def f(*args2: P2.args, **kwargs2: P2.kwargs) -> None:
+            reveal_type(join(args, args2))  # N: Revealed type is "builtins.object*"
+            reveal_type(join(kwargs, kwargs2))  # N: Revealed type is "builtins.object*"
+
+    def m3(self, c: C[P, P3]) -> None:
+        reveal_type(join(c, c))  # N: Revealed type is "__main__.C*[P`1, P3`-1]"
+        reveal_type(join(self, c))  # N: Revealed type is "builtins.object*"
+[builtins fixtures/dict.pyi]
+
+[case testParamSpecClassWithAny]
+from typing import Callable, Generic, Any
+from typing_extensions import ParamSpec
+
+P = ParamSpec('P')
+
+class C(Generic[P]):
+    def __init__(self, x: Callable[P, None]) -> None: ...
+
+    def m(self, *args: P.args, **kwargs: P.kwargs) -> int:
+        return 1
+
+c: C[Any]
```

Comment:
Would any other non-`Any` type arguments work? I haven't see any examples of it in tests 🤔

Reply:
This is not implemented. I've mentioned it in the commit summary as follow-up work.